### DISCUSSION TITLE
WRN Authentication rejected: the OpenID token was rejected: the OpenId token is signed with an unknown key

### DISCUSSION THREAD

**Discussion by sharadkhoji123** [kiali/kiali#7471] [General]
I tried using Azure AD authentication with OpenID in Kiali. 

Below are the entries I have made:
```data:
  config.yaml: |
    additional_display_details:
    - annotation: kiali.io/api-spec
      icon_annotation: kiali.io/api-type
      title: API Documentation
    auth:
      strategy: openid
      openid:
        api_token: "access_token"
        authentication_timeout: 300
        authorization_endpoint: "https://login.microsoftonline.com/************************/oauth2/authorize"
        token_endpoint: "https://login.microsoftonline.com/**********************/oauth2/token"
        client_id: "*****************************"
        disable_rbac: true
        insecure_skip_verify_tls: false
        issuer_uri: "https://login.microsoftonline.com/*************************/v2.0"
        scopes:
        - openid
        - profile
        - email
        username_claim: "email"
        password: secret:kiali:oidc-secret
```
Secret is already present in the environment:
k get secret kiali -n istio-system
NAME    TYPE     DATA   AGE
kiali   Opaque   1      8h
I have registered redirect uri as "https://isitio-traffik.example.com/kiali" in the Azure App. Also I have created roleBinding with kiali and kiali-viewer roles to my email id which is part of Azure App.
Below are the binding definitions:
```apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: kiali-viewer-binding
subjects:
- kind: User
  name: s****l@test.com
  apiGroup: rbac.authorization.k8s.io
roleRef:
  kind: ClusterRole
  name: kiali-viewer
  apiGroup: rbac.authorization.k8s.io
...
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: kiali-binding
subjects:
- kind: User
  name: s****l@test.com
  apiGroup: rbac.authorization.k8s.io
roleRef:
  kind: ClusterRole
  name: kiali
  apiGroup: rbac.authorization.k8s.io
```
When I try to login I get below error in kiali pod log:
2024-06-25T17:36:17Z INF Kiali: Version: v1.82.1, Commit: 539fc95c5a5311c3efbbf338a69e84fd9268daad, Go: 1.20.10
2024-06-25T17:36:17Z INF Using authentication strategy [openid]
2024-06-25T17:36:17Z INF Some validation errors will be ignored [KIA1301]. If these errors do occur, they will still be logged. If you think the validation errors you see are incorrect, please report them to the Kiali team if you have not done so already and provide the details of your scenario. This will keep Kiali validations strong for the whole community.
2024-06-25T17:36:17Z INF Initializing Kiali Cache
2024-06-25T17:36:17Z INF [Kiali Cache] Waiting for cluster-scoped cache to sync
2024-06-25T17:36:17Z INF [Kiali Cache] Started
2024-06-25T17:36:17Z INF [Kiali Cache] Kube cache is active for cluster: [Kubernetes]
2024-06-25T17:36:18Z INF [Prom Cache] Enabled
2024-06-25T17:36:18Z INF Server endpoint will start at [:20001/kiali]
2024-06-25T17:36:18Z INF Server endpoint will serve static content from [/opt/kiali/console]
2024-06-25T17:36:18Z INF Starting Metrics Server on [:9090]
2024-06-25T17:37:16Z WRN Authentication rejected: the OpenID token was rejected: the OpenId token is signed with an unknown key
2024-06-25T17:37:28Z WRN Authentication rejected: the OpenID token was rejected: the OpenId token is signed with an unknown key

Kindly help I have already reviewed discussion "https://github.com/kiali/kiali/discussions/4243" but that is not helping me. Kindly help thanks in advance.

**Comment by jmazzitelli** [comment]
I don't have an answer, but I do want to point out that to correctly format code or yaml in github comments so people can see your indented output and code easily, wrap it in triple quotes, please. For yaml, you can even use "````yaml" so github color-formats. E.g.

  \```yaml
    your yaml here
  \```

results in something like this:
```yaml
outer:
  inner: foo
```

**Reply by sharadkhoji123** [reply]
@jmazzitelli I have updated yaml within quotes

**Comment by nrfox** [comment]
From the error it looks like something might be wrong with your oidc secret. Can you double check you have created it properly according to: https://kiali.io/docs/configuration/authentication/openid/#setup-with-rbac? I see `password: secret:kiali:oidc-secret` in your configuration but this is not correct. It needs to be a secret named `kiali` with a key named `oidc-secret` in the same namespace as kiali.

**Reply by sharadkhoji123** [reply]
Hi @nrfox thanks for the response.
I have already followed the approach as mentioned in the link without passing any client secret and just maintaining a secret called kiali with key oidc-secret, but this did not work either. 
Also I tried passing it client_secret direct as key under auth tree but that also did not work. 
The one way I am following currently ```password: secret:kiali:oidc-secret``` is the one which worked for ssathyak in https://github.com/kiali/kiali/discussions/4243 discussion.

**Reply by jmazzitelli** [reply]
FWIW: there is no "password" setting in auth.openid... that is unused/ignored. Please see: 
* https://kiali.io/docs/configuration/kialis.kiali.io/#.spec.auth.openid
* https://github.com/kiali/kiali/blob/v1.86.0/config/config.go#L407-L424

I haven't looked deep yet, but that secret is assumed to be mounted on the file system at /kiali-secret/oidc-secret: https://github.com/kiali/kiali/blob/master/config/config.go#L52

Somehow the operator must know to mount that to the server pod.

But my main point is - that password field is irrelevent AFAICS.

**Reply by jmazzitelli** [reply]
OK, I see how its done. That's this setting: https://kiali.io/docs/configuration/kialis.kiali.io/#.spec.deployment.secret_name

Read the description. That is how the operator knows to mount the secret to the pod (here's where it does that: https://github.com/kiali/kiali-operator/blob/v1.86.0/roles/default/kiali-deploy/templates/kubernetes/deployment.yaml#L117-L118 -- and here https://github.com/kiali/kiali-operator/blob/v1.86.0/roles/default/kiali-deploy/templates/kubernetes/deployment.yaml#L145-L148)
